source/server/hot_restart_impl.cc

```diff
@@ -156,7 +164,8 @@ void HotRestartImpl::free(Stats::RawStatData& data) {
   if (--data.ref_count_ > 0) {
     return;
   }
-
+  bool key_removed = stats_set_.remove(data.key());
+  RELEASE_ASSERT(key_removed);
```

Comment:
nit: I would probably make this a normal ASSERT

Reply:
That will fail to compile in release builds, because key_removed is assigned and not used when the ASSERT disappears (found that through CI but it's a pattern I've seen many times).

The pattern we'd use at Google for this is:

```
 bool key_removed = stats_set_.remove(data.key());
 if (!key_removed) {
   LOG(DFATAL) << "key not removed for key " << data.key();
}
```

In my vanilla C++ life I'd do:

```
#ifndef NDEBUG
assert(stats_set_.remove(data.key()));
#else 
stats_set_.remove(data.key());
#endif
```

I'd be happy with whatever you think makes the most sense in the Envoy codebase.

Reply:
In these cases in the code in Envoy we generally do `UNREFERENCED_PARAMETER(key_removed)`

Reply:
Still need to switch to `ASSERT` :)

Reply:
sorry, fixed now :)  Also fixed a broken integration test due to the way SharedMemory::version() was called directly from main().  Made it call a new HotRestart::hotRestartVersion() method which also pulls in the SharedMemoryHashSet data.